### PR TITLE
Revoker Overhaul

### PR DESCRIPTION
**Pull request by ylv-io**
# Description

This PR is a part of #2242 issue. It is an overhaul of `revoker` functionality in `TimelockAuthorizer`.

It changes addRevoker, removeRevoker, isRevoker, and canRevoke so that revoker permissions are stored in a separate data structure, and are not regular permissions (i.e. they don't use the 'base permission' and 'specifier' scheme). This makes analyzing the rules for granting and revoking permissions much simpler. Revoking now also has dedicated delays.

The API is the following:
* addRevoker
* removeRevoker
* isRevoker
* canRevoke (This is isRevoker + delay semantics)
* scheduleRevokePermission

There are two changes to the behavior of the code:
* Root is always a revoker of everything everywhere (and this cannot be undone).
* Since root can already revoke everything, I removed the notion of being a 'general revoker' (a revoker of any permission), since I find no reasonable use case for this. We can still have global revokers of a permission (e.g. can revoke permission to set swap fees on any pool), but not of all permissions.

I modified addRevoker/removeRevoker tests to reflect changes. I skipped the tests that are related to delays in revoking, waiting on @nventuro  to implement this and would be a separate PR. 

Tests are a complete overhaul so I recommend to ignore the diff and just to review the new tests under `revokers` block.

## Type of change

- [ ] Bug fix 
- [ ] New feature 
- [x] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- [x] Complex code has been commented, including external interfaces
- [x] Tests are included for all code paths
- [x] The base branch is either `master`, or there's a description of how to merge

### CODE REVIEW

File: pkg/vault/contracts/authorizer/TimelockAuthorizer.sol [hunk 1]
```diff
@@ -155,6 +159,16 @@ contract TimelockAuthorizer is IAuthorizer, IAuthentication, ReentrancyGuard {
      */
     event GranterRemoved(bytes32 indexed actionId, address indexed account, address indexed where);
 
+    /**
+     * @notice Emitted when an account is added as a revoker for `actionId` in `where`.
```
Comment (EndymionJkb) [inline pkg/vault/contracts/authorizer/TimelockAuthorizer.sol:163]:
```suggestion
     * @notice Emitted when `account` is added as a revoker for `actionId` in `where`.
```
Complete nit

File: pkg/vault/contracts/authorizer/TimelockAuthorizer.sol [hunk 2]
```diff
@@ -850,8 +871,20 @@ contract TimelockAuthorizer is IAuthorizer, IAuthentication, ReentrancyGuard {
...
+        require(isRevoker(actionId, account, where), "ACCOUNT_IS_NOT_REVOKER");
+
+        require(!isRoot(account), "CANNOT_REMOVE_ROOT_REVOKER");
+
+        // On top of requiring that the account is currently a revoker, we prevent attempts to remove permission over a
+        // single contract from global revokers. As mentioned in `addRevoker`, it is possible for an account to have
+        // both global and specific permissions over a given contract: in this case, the global permission must be
+        // removed before the specific ones can be addressed.
+        if (_isRevoker[actionId][account][EVERYWHERE]) {
+            require(where == EVERYWHERE, "REVOKER_IS_GLOBAL");
```
Comment (EndymionJkb) [inline pkg/vault/contracts/authorizer/TimelockAuthorizer.sol:883]: Maybe "ACCOUNT_IS_GLOBAL_REVOKER" ?

File: pkg/vault/contracts/authorizer/TimelockAuthorizer.sol [hunk 3]
```diff
@@ -884,9 +917,14 @@ contract TimelockAuthorizer is IAuthorizer, IAuthentication, ReentrancyGuard {
         address[] memory executors
     ) external returns (uint256 scheduledExecutionId) {
         require(isRevoker(actionId, msg.sender, where), "SENDER_IS_NOT_REVOKER");
+
+        uint256 delay = _revokeDelays[actionId];
+        require(delay > 0, "ACTION_HAS_NO_GRANT_DELAY");
```
Comment (nventuro) [inline pkg/vault/contracts/authorizer/TimelockAuthorizer.sol:922]:
```suggestion
        require(delay > 0, "ACTION_HAS_NO_REVOKE_DELAY");
```

:upside_down_face:

**Review by nventuro — APPROVED** [review]
Fantastic :) Thanks a lot @ylv-io!

File: pkg/vault/contracts/authorizer/TimelockAuthorizer.sol (hunk 2 above)
Comment (ylv-io) [reply pkg/vault/contracts/authorizer/TimelockAuthorizer.sol:883]: I've picked this for consistency with `require(where == EVERYWHERE, "GRANTER_IS_GLOBAL");`. I think `ACCOUNT_IS_GLOBAL_REVOKER` is better but I would change both of them in a separate PR.

**Pull request merged**